# Bellabeat Case Study

# 1\. Defining A Problem

Bellabeat is a successful small company, but they have the potential to become a larger player in the global smart device market. Urška Sršen, cofounder and Chief Creative Officer of Bellabeat, believes that analyzing smart device fitness data could help unlock new growth opportunities for the company. You have been asked to focus on one of Bellabeat’s products and analyze smart device data to gain insight into how consumers are using their smart devices. The insights you discover will then help guide marketing strategy for the company. You will present your analysis to the Bellabeat executive team along with your high-level recommendations for Bellabeat’s marketing strategy.

### **Characters:**

- **Urška Sršen:** Bellabeat’s cofounder and Chief Creative Officer 
- **Sando Mur:** Mathematician and Bellabeat’s cofounder; key member of the Bellabeat executive team 
- **Bellabeat marketing analytics team:** A team of data analysts responsible for collecting, analyzing, and reporting data that helps guide Bellabeat’s marketing strategy. You joined this team six months ago and have been busy learning about Bellabeat’’s mission and business goals — as well as how you, as a junior data analyst, can help Bellabeat achieve them.

**A business task is defined by the next questions:**

1. What are some trends in smart device usage? 
2. How could these trends apply to Bellabeat customers? 
3. How could these trends help influence Bellabeat marketing strategy?

# 2\. Preparing Data

To solve the problem I am given a database 'bellabeat', which includes datasets:

1\. 'dailyActivity' (940 records, 15 fields), which describes users' daily activity statistics, such as daily distances walked, steps made, activity level times and distances, and calories burned:

In [8]:
SELECT COUNT(*) AS number_of_rows FROM dailyActivity;
SELECT TOP 5 * FROM dailyActivity;

(1 row affected)

(5 rows affected)

Total execution time: 00:00:00.068

number_of_rows 940

Id ActivityDate TotalSteps TotalDistance TrackerDistance LoggedActivitiesDistance VeryActiveDistance ModeratelyActiveDistance LightActiveDistance SedentaryActiveDistance VeryActiveMinutes FairlyActiveMinutes LightlyActiveMinutes SedentaryMinutes Calories 1503960366 4/12/2016 13162 8.5 8.5 0 1.87999999523163 0.550000011920929 6.05999994277954 0 25 13 328 728 1985 1503960366 4/13/2016 10735 6.96999979019165 6.96999979019165 0 1.57000005245209 0.689999997615814 4.71000003814697 0 21 19 217 776 1797 1503960366 4/14/2016 10460 6.73999977111816 6.73999977111816 0 2.44000005722046 0.400000005960464 3.91000008583069 0 30 11 181 1218 1776 1503960366 4/15/2016 9762 6.28000020980835 6.28000020980835 0 2.14000010490417 1.25999999046326 2.82999992370605 0 29 34 209 726 1745 1503960366 4/16/2016 12669 8.15999984741211 8.15999984741211 0 2.71000003814697 0.409999996423721 5.03999996185303 0 36 10 221 773 1863

2\. 'heartrateSeconds' (2483658 records, 3 fields) - shows all the users' heartrates recorded by the app, including user's ID, record date-time and the heartrate value:

In [7]:
SELECT COUNT(*) AS number_of_rows FROM heartrateSeconds;
SELECT TOP 5 * FROM heartrateSeconds;

(1 row affected)

(5 rows affected)

Total execution time: 00:00:00.774

number_of_rows 2483658

Id Time Value 2022484408 4/30/2016 1:16:15 PM 77 2022484408 4/30/2016 1:16:20 PM 79 2022484408 4/30/2016 1:16:25 PM 81 2022484408 4/30/2016 1:16:30 PM 80 2022484408 4/30/2016 1:16:40 PM 79

3\. 'sleepDay' (413 records, 5 fields) - about daily sleep time and time in bed of each user.

In [9]:
SELECT COUNT(*) AS numberOfRows FROM sleepDay;
SELECT TOP 5 * FROM sleepDay;

(1 row affected)

(5 rows affected)

Total execution time: 00:00:00.042

numberOfRows 413

Id SleepDay TotalSleepRecords TotalMinutesAsleep TotalTimeInBed 1503960366 4/12/2016 12:00:00 AM 1 327 346 1503960366 4/13/2016 12:00:00 AM 2 384 407 1503960366 4/15/2016 12:00:00 AM 1 412 442 1503960366 4/16/2016 12:00:00 AM 2 340 367 1503960366 4/17/2016 12:00:00 AM 1 700 712

4\. weightLogInfo (67 rows, 8 fields) - the data about users' weights and body mass indexes.

In [10]:
SELECT COUNT(*) AS numberOfRows FROM weightLogInfo;
SELECT TOP 5 * FROM weightLogInfo;

(1 row affected)

(5 rows affected)

Total execution time: 00:00:00.044

numberOfRows 67

Id Date WeightKg WeightPounds Fat BMI IsManualReport LogId 1503960366 5/2/2016 11:59:59 PM 52.5999984741211 115.963146545323 22 22.6499996185303 True 1462233599000 1503960366 5/3/2016 11:59:59 PM 52.5999984741211 115.963146545323 22.6499996185303 True 1462319999000 1927972279 4/13/2016 1:08:52 AM 133.5 294.317120016975 47.5400009155273 False 1460509732000 2873212765 4/21/2016 11:59:59 PM 56.7000007629395 125.002104340889 21.4500007629395 True 1461283199000 2873212765 5/12/2016 11:59:59 PM 57.2999992370605 126.324874550011 21.6900005340576 True 1463097599000

Then I used to check if all datasets are full in term of sample size (30 Bellabeat users should be included into testing and statistics):

In [11]:
SELECT COUNT(DISTINCT Id) AS dailyActivityIDs FROM dailyActivity;
SELECT COUNT(DISTINCT Id) AS heartrateIDs     FROM heartrateSeconds;
SELECT COUNT(DISTINCT Id) AS sleepDayIDs      FROM sleepDay;
SELECT COUNT(DISTINCT Id) AS weightLogIDs     FROM weightLogInfo;

(1 row affected)

(1 row affected)

(1 row affected)

(1 row affected)

Total execution time: 00:00:01.470

dailyActivityIDs 33

heartrateIDs 14

sleepDayIDs 24

weightLogIDs 8

Number of users in each dataset:

dailyActivity - 33

heartrateSeconds - 14

sleepDayIDs - 24

weightLogInfo - 8

Only 'dailyActivity' is full in terms of number of users (30+), so the analysis can be unfair.

  

# 3\. Cleaning

## 3.1. Cleaning 'dailyActivity'

### 3.1.1. Checking for duplicates

In [1]:
SELECT Id, ActivityDate, COUNT(*) AS numberOfRecords
FROM dailyActivity
GROUP BY Id, ActivityDate
HAVING COUNT(*) > 1

(0 rows affected)

Total execution time: 00:00:00.267

Id ActivityDate numberOfRecords

No duplicates.

3.1.2. Removing unuseful columns 

'SedentaryActivrDistance' column if wrong, because a human cannotwalk while being sedentary. LoggedActivitiesDistance is also unnesessary until almost all values in this column equal to 0.

In [2]:
ALTER TABLE dailyActivity 
DROP COLUMN SedentaryActiveDistance, LoggedActivitiesDistance;

Commands completed successfully.

Total execution time: 00:00:00.109

3.1.3. Rename column 'ActivityDate' to 'Date'

I used to do this for column name consistency between tables.

In [3]:
EXEC sp_rename 'dailyActivity.ActivityDate', 'Date', 'COLUMN';

Caution: Changing any part of an object name could break scripts and stored procedures.

Total execution time: 00:00:00.519

3.1.4. Checking datatypes

In [7]:
EXEC sp_help dailyActivity;

The object 'dailyActivity' does not have any indexes, or you do not have permissions.

No constraints are defined on object 'dailyActivity', or you do not have permissions.

No foreign keys reference table 'dailyActivity', or you do not have permissions on referencing tables.

No views with schema binding reference table 'dailyActivity'.

Total execution time: 00:00:00.633

Name Owner Type Created_datetime dailyActivity dbo user table 2022-07-23 19:39:25.867

Column_name Type Computed Length Prec Scale Nullable TrimTrailingBlanks FixedLenNullInSource Collation Id varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS Date varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS TotalSteps varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS TotalDistance varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS TrackerDistance varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS VeryActiveDistance varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS ModeratelyActiveDistance varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS LightActiveDistance varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS VeryActiveMinutes varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS FairlyActiveMinutes varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS LightlyActiveMinutes varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS SedentaryMinutes varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS Calories varchar no 50 yes no yes SQL_Latin1_General_CP1_CI_AS

Identity Seed Increment Not For Replication No identity column defined. NULL NULL NULL

RowGuidCol No rowguidcol column defined.

Data_located_on_filegroup PRIMARY

All the columns have varchar datatype, and that is incorrect. No columns should have varchar datatype:

\- 'Id', 'TotalSteps', 'VeryActiveMinutes', 'FairlyActiveMinutes', 'LightlyActiveMinutes', 'SedentaryMinutes' and 'Calories' - integer;

\- 'Date' - Date datatype;

\- 'TotalDistance', 'TrackerDistance', 'VeryActiveDistance', 'ModeratelyActiveDistance', 'LightActiveDistance' - double precision,

In [32]:
ALTER TABLE dailyActivity ALTER COLUMN Id bigint;
ALTER TABLE dailyActivity ALTER COLUMN Date DATE;
ALTER TABLE dailyActivity ALTER COLUMN TotalSteps INTEGER;
ALTER TABLE dailyActivity ALTER COLUMN TotalDistance FLOAT; 
ALTER TABLE dailyActivity ALTER COLUMN TrackerDistance FLOAT;
ALTER TABLE dailyActivity ALTER COLUMN VeryActiveDistance FLOAT; 
ALTER TABLE dailyActivity ALTER COLUMN ModeratelyActiveDistance FLOAT;
ALTER TABLE dailyActivity ALTER COLUMN LightActiveDistance FLOAT;
ALTER TABLE dailyActivity ALTER COLUMN VeryActiveMinutes INTEGER;
ALTER TABLE dailyActivity ALTER COLUMN FairlyActiveMinutes INTEGER;
ALTER TABLE dailyActivity ALTER COLUMN LightlyActiveMinutes INTEGER;
ALTER TABLE dailyActivity ALTER COLUMN SedentaryMinutes INTEGER;
ALTER TABLE dailyActivity ALTER COLUMN Calories INTEGER;

Commands completed successfully.

Total execution time: 00:00:00.512

3.1.5. Adding variable 'CaloriesLevel' 

This column will decribe the level of calories burned:

\- if Calories \< 2000, the CaloriesLevel = Low;

\- if Calories \>= 2000, then the CaloriesLevel = Normal.

In [34]:
ALTER TABLE dailyActivity ADD CaloriesLevel AS (CASE WHEN Calories < 2000 THEN 'Low' ELSE 'Normal' END);

Commands completed successfully.

Total execution time: 00:00:00.057

### 3.1.6. Checking for ouliers

In terms of mathematical statistic, outlier is a value that is not in range (average value +- 3\* standard deviation), so I used to check every variable for having outliers:

In [29]:
-- checking the number of records by date
WITH records AS 
(
    SELECT
        Date,
        COUNT(*) OVER (PARTITION BY Date ORDER BY Date) AS NumberOfRecords -- counting num of records made each day
    FROM dailyActivity
),
stats AS
(
    SELECT 
        STDEV(NumberOfRecords) AS StDeviation, -- counting standard deviation 
        AVG(NumberOfRecords) AS Average -- counting average value
    FROM records
)
-- defining a date outlier 
SELECT DISTINCT Date, NumberOfRecords
FROM records, stats
WHERE NumberOfRecords < Average - 3 * StDeviation OR NumberOfRecords > Average + 3 * StDeviation;
-- 2016-05-12 is an outlier with far less records made by users, because it was the last day of tracking

(1 row affected)

Total execution time: 00:00:00.083

Date NumberOfRecords 2016-05-12 21

In [32]:
-- checking the steps made each day
WITH stats AS
(
    SELECT 
        STDEV(TotalSteps) AS StDeviation, -- counting standard deviation 
        AVG(TotalSteps) AS Average -- counting average value
    FROM dailyActivity
)
-- defining a date outlier 
SELECT DISTINCT Id, Date, TotalSteps
FROM dailyActivity, stats
WHERE TotalSteps < Average - 3 * StDeviation OR TotalSteps > Average + 3 * StDeviation;
-- there are six outliers than need some additional analysis to define what to do with them

(6 rows affected)

Total execution time: 00:00:00.037

Id Date TotalSteps 1624580081 2016-05-01 36019 8053475328 2016-04-24 22988 8877689391 2016-04-12 23186 8877689391 2016-04-16 29326 8877689391 2016-04-27 23629 8877689391 2016-04-30 27745

In [34]:
-- checking the calories burned by users each date
WITH stats AS
(
    SELECT 
        STDEV(Calories) AS StDeviation, -- counting standard deviation 
        AVG(Calories) AS Average -- counting average value
    FROM dailyActivity
)
-- defining a date outlier 
SELECT DISTINCT Id, Date, Calories
FROM dailyActivity, stats
WHERE Calories < Average - 3 * StDeviation OR Calories > Average + 3 * StDeviation
ORDER BY Date;
-- there are 12 outliers 

(12 rows affected)

Total execution time: 00:00:00.062

Id Date Calories 8877689391 2016-04-16 4547 5577150313 2016-04-17 4552 6117666160 2016-04-21 4900 5577150313 2016-04-30 4501 8253242879 2016-04-30 0 5577150313 2016-05-01 4546 7007744171 2016-05-07 120 6290855005 2016-05-10 0 8792009665 2016-05-10 57 3977333714 2016-05-11 52 1503960366 2016-05-12 0 8583815059 2016-05-12 0

Well, according to https://www.sleepfoundation.org/how-sleep-works/how-your-body-uses-calories-while-you-sleep#:~:text=How%20Many%20Calories%20Do%20You,metabolic%20rate2%20(BMR).

Human burns at least 50 calories per hour even when doing nothing, sleeping and being sedentary. That's 1200 calories per day, so there's no way to burn less. I used to remove all the numbers less than 1200.

In [35]:
DELETE FROM dailyActivity WHERE Calories < 1200;

(18 rows affected)

Total execution time: 00:00:00.091

## 3.2. Cleaning 'heartrateSeconds' dataset

3.2.1. Changing datatypes

  

All the data in the table has 'varchar' data type, so we need to change it to the types corresponding to the data:

\- 'Id' - bigint;

\- 'Time' - datetime;

\- 'Value' - integer.

In [37]:
ALTER TABLE heartrateSeconds ALTER COLUMN Id bigint;
ALTER TABLE heartrateSeconds ALTER COLUMN Time datetime;
ALTER TABLE heartrateSeconds ALTER COLUMN Value INTEGER;

Commands completed successfully.

Total execution time: 00:04:52.515

3.2.2. Renaming column 'Value' to 'Heartrate'

In [38]:
EXEC sp_rename 'heartrateSeconds.Value', 'Heartrate', 'COLUMN';

Caution: Changing any part of an object name could break scripts and stored procedures.

Total execution time: 00:00:00.342

3.2.3. Adding variables 'Date', 'Hour' and 'Minute'

In [48]:
ALTER TABLE heartrateSeconds ADD 
    Date AS (CAST(Time AS date)),
    Hour AS DATEPART(HOUR, Time),
    Minute AS DATEPART(MINUTE, Time);

Commands completed successfully.

Total execution time: 00:00:00.003

3.2.4. Modifying table

There's no need to save multiple heartrate records per minute, so I used to round 'Time' to minutes

In [58]:
ALTER TABLE heartrateSeconds ADD Datetime AS FORMAT(Time, 'dd-MM-yyyy HH:mm') 

Commands completed successfully.

Total execution time: 00:00:00.004

Then, rather than maintain a table with multiple heartrates per minute, I used to create table with average heartrates per minute, it would contain 10x less records and memory.

In [64]:
SELECT
    Id,
    Datetime,
    Date,
    Hour,
    CAST(AVG(Heartrate) AS int) AS HeartRate
INTO heartrateMinutes
FROM heartrateSeconds
GROUP BY Id, Datetime, Date, Hour;

(333420 rows affected)

Total execution time: 00:01:36.531

### 3.2.5. Adding variable 'HeartrateLevel'

There are three levels of human's herartrate:

\- Low - less than 60 beats per minute (bpm);

\- Normal - 60-100 bpm;

\- High - more than 100 bpm.

  

I will include this information into the table:

In [65]:
ALTER TABLE heartrateMinutes ADD HeartrateLevel AS
(
    CASE 
        WHEN HeartRate < 60 THEN 'Low'     -- if a heartrate is lower than 60 bpm, then that's a low heartrate
        WHEN HeartRate > 100 THEN 'High'   -- if a heartrate is higher than 100 bpm, then that's a high heartrate
        ELSE 'Normal'                      -- if a heartrate is 60-100 bpm, then that's a normal heartrate
    END
);

Commands completed successfully.

Total execution time: 00:00:00.020

Then I used to delete 'heartrateSeconds' dataset to save memory:

In [66]:
DROP TABLE heartrateSeconds;

Commands completed successfully.

Total execution time: 00:00:00.031

### 3.2.6. Checking for duplicates

In [69]:
SELECT
    Id, 
    Datetime,
    COUNT(*) AS numOfDuplicates
FROM heartrateMinutes
GROUP BY Id, Datetime
HAVING COUNT(*) > 1; -- checking rows that have multiple entries = duplicates

(0 rows affected)

Total execution time: 00:00:01.228

Id Datetime numOfDuplicates

## 3.3. Processing 'sleepDay' dataset

### 3.3.1. Changing datatypes

In [71]:
ALTER TABLE sleepDay ALTER COLUMN Id bigint; -- changing datatype of Id from varchar to big integer
ALTER TABLE sleepDay ALTER COLUMN SleepDay Date; -- changing datatype from varchar to date
ALTER TABLE sleepDay ALTER COLUMN TotalSleepRecords int; -- from varchar to int
ALTER TABLE sleepDay ALTER COLUMN TotalMinutesAsleep int; -- from varchar to int
ALTER TABLE sleepDay ALTER COLUMN TotalTimeInBed int; -- from varchar to int

Commands completed successfully.

Total execution time: 00:00:00.111

### 3.3.2. Checking for duplicates

In [72]:
SELECT 
    Id,
    SleepDay,
    COUNT(*) AS NumOfDuplicates
FROM sleepDay
GROUP BY Id, SleepDay
HAVING COUNT(*) > 1; -- detect records that are repeated in the table (duplicates)

(3 rows affected)

Total execution time: 00:00:00.045

Id SleepDay NumOfDuplicates 8378563200 2016-04-25 2 4388161847 2016-05-05 2 4702921684 2016-05-07 2

There are 3 duplicate entries, so I must delete them.

In [73]:
WITH duplicateTable AS 
(
    SELECT *, 
        ROW_NUMBER() OVER -- count the number of unique entries
        (
            PARTITION BY Id, SleepDay -- group by primary key (Id + SleepDay) but not change the number of rows
            ORDER BY Id, SleepDay -- sort by primary key
        ) AS numOfRecords
    FROM sleepDay
)
DELETE FROM duplicateTable WHERE numOfRecords <> 1; -- delete all the non-unique entries with keeping the original

(3 rows affected)

Total execution time: 00:00:00.033

### 3.3.3. Rename 'SleepDay' column to 'Date'

In [76]:
EXEC sp_rename 'sleepDay.SleepDay', 'Date', 'COLUMN';

Caution: Changing any part of an object name could break scripts and stored procedures.

Total execution time: 00:00:00.549

### 3.3.4. Adding 'SleepRate' column

Daily sleeping norm for women is 7-9 hours (420-540 minutes) . that's normal level of sleeping. 

If less than 7 hours - 'Lack of dleep',

If more than 9 hours - 'Oversleeping'

In [77]:
ALTER TABLE sleepDay 
ADD SleepRate AS
(
    CASE 
    WHEN TotalMinutesAsleep < 420 THEN 'Lack of sleep'
    WHEN TotalMinutesAsleep > 540 THEN 'Oversleeping'
    ELSE 'Normal sleep'
    END
);

Commands completed successfully.

Total execution time: 00:00:00.069

### 3.3.5. Checking for outliers

In [38]:
-- checking the number of records by date
WITH stats AS
(
    SELECT 
        STDEV(TotalMinutesAsleep) AS StDeviation, -- counting standard deviation 
        AVG(TotalMinutesAsleep) AS Average -- counting average value
    FROM sleepDay
)
-- defining a date outlier 
SELECT DISTINCT Id, Date, TotalMinutesAsleep
FROM sleepDay, stats
WHERE TotalMinutesAsleep < Average - 3 * StDeviation OR TotalMinutesAsleep > Average + 3 * StDeviation;
-- 6 outliers than need an additional analysis

(6 rows affected)

Total execution time: 00:00:00.107

Id Date TotalMinutesAsleep 1644430081 2016-05-02 796 2320127002 2016-04-23 61 4319703577 2016-04-21 59 4388161847 2016-05-09 62 5553957443 2016-04-30 775 7007744171 2016-05-01 58

### 3.3.6. Adding 'TimeBeforeSleep' column

Which describes how much time we spend from going in bed and finally falling asleep.

In [39]:
ALTER TABLE sleepDay ADD TimeBeforeSleep AS (TotalTimeInBed - TotalMinutesAsleep);

Commands completed successfully.

Total execution time: 00:00:00.053

In [40]:
SELECT TOP 5 * FROM sleepDay;

(5 rows affected)

Total execution time: 00:00:00.027

Id Date TotalSleepRecords TotalMinutesAsleep TotalTimeInBed SleepRate TimeBeforeSleep 1503960366 2016-04-12 1 327 346 Lack of sleep 19 1503960366 2016-04-13 2 384 407 Lack of sleep 23 1503960366 2016-04-15 1 412 442 Lack of sleep 30 1503960366 2016-04-16 2 340 367 Lack of sleep 27 1503960366 2016-04-17 1 700 712 Oversleeping 12

## 3.4. Cleaning 'weightLogInfo'

### 3.4.1. Removing 'FAT' column

In [82]:
SELECT COUNT(*) AS NumberOfNAs FROM weightLogInfo 
WHERE FAT = ''; 

(1 row affected)

Total execution time: 00:00:00.018

NumberOfNAs 65

65/67 values of FAT column are empty, so the data in this column is incomplete, and I must remove the column.

In [3]:
ALTER TABLE weightLogInfo DROP COLUMN FAT;

Commands completed successfully.

Total execution time: 00:00:00.064

### 3.4.2. Changing data types of all columns

In [4]:
ALTER TABLE weightLogInfo ALTER COLUMN Id bigint; -- change 'Id' data type from varchar to big integer
ALTER TABLE weightLogInfo ALTER COLUMN Date date; -- change 'Date' data type from varchar to date
ALTER TABLE weightLogInfo ALTER COLUMN WeightKg float; -- change 'WeightKg' data type from varchar to float
ALTER TABLE weightLogInfo ALTER COLUMN WeightPounds float; -- change 'WeightPounds' data type from varchar to float
ALTER TABLE weightLogInfo ALTER COLUMN BMI float; -- change 'BMI' data type from varchar to float
ALTER TABLE weightLogInfo ALTER COLUMN IsManualReport bit; -- change 'IsManualReport' data type from varchar to bit
ALTER TABLE weightLogInfo ALTER COLUMN LogId bigint; -- change 'LogId' data type from varchar to big integer

Commands completed successfully.

Total execution time: 00:00:00.089

### 3.4.3. Checking for duplicates

In [6]:
SELECT 
    Id,
    Date,
    COUNT(*) AS NumOfDuplicates -- count the number of primary key (Id + Date) entries
FROM weightLogInfo
GROUP BY Id, Date
HAVING COUNT(*) > 1; -- output all the non-unique values

(0 rows affected)

Total execution time: 00:00:00.016

Id Date NumOfDuplicates

There's no duplicates.

### 3.4.4. Rename 'BMI' to 'BodyMassIndex'

BMI is an abbreviation, so not everybody can understand it. I used to name the column by its full name.

In [7]:
-- rename weightLogInfo column from BMI to BodyMassIndex using sp_rename function
EXEC sp_rename 'weightLogInfo.BMI', 'BodyMassIndex', 'COLUMN';

Caution: Changing any part of an object name could break scripts and stored procedures.

Total execution time: 00:00:00.538

### 3.4.4. Adding 'BodyFatRank' column

This column will rank womens' BMI according to the next criterias:

- 'Normal' BMI - 18.5-24.9 kg/m^2;
- 'Underweight' - less than 18.5 kg/m^2;
- 'Overweight' - more than 24.9 kg/m^2.

In [8]:
ALTER TABLE weightLogInfo ADD BodyFatRank AS -- add column 'BodyFatRank' which is calculated as:
(
    CASE
    WHEN BodyMassIndex < 18.5 THEN 'Underweight' -- if a BMI is lower than 18.4, then a person is underweight
    WHEN BodyMassIndex > 24.9 THEN 'Overweight' -- if a BMI is higher than 24.9, then a person is overweight
    ELSE 'Normal' -- if a BMI is 18.5-24.9, the person has normal body fat
    END
);

Commands completed successfully.

Total execution time: 00:00:00.013

# 4\. Analysis

## 4.1. Finding the start and end date of tracking in each dataset

In [10]:
SELECT MIN(Date) AS StartDate, MAX(Date) AS EndDate FROM dailyActivity;
SELECT MIN(Date) AS StartDate, MAX(Date) AS EndDate FROM sleepDay;
SELECT MIN(Date) AS StartDate, MAX(Date) AS EndDate FROM heartrateMinutes;
SELECT MIN(Date) AS StartDate, MAX(Date) AS EndDate FROM weightLogInfo;
-- in all tables tracking time is from 2016-05-12 to 2016-05-12

(1 row affected)

(1 row affected)

(1 row affected)

(1 row affected)

Total execution time: 00:00:00.234

StartDate EndDate 2016-04-12 2016-05-12

StartDate EndDate 2016-04-12 2016-05-12

StartDate EndDate 2016-04-12 2016-05-12

StartDate EndDate 2016-04-12 2016-05-12

## 4.2. 'dailyActivity' explorating

### Researching averages

In [41]:
-- average scores of each Id
SELECT
    Id,
    AVG(TotalSteps) AS AvgSteps,
    AVG(VeryActiveDistance) AS AvgVeryActiveDistance,
    AVG(VeryActiveMinutes) AS AvgVeryActiveMinutes,
    AVG(SedentaryMinutes) AS AvgSedentary,
    AVG(Calories) AS AvgCalories
FROM dailyActivity
GROUP BY Id;

(33 rows affected)

Total execution time: 00:00:00.093

Id AvgSteps AvgVeryActiveDistance AvgVeryActiveMinutes AvgSedentary AvgCalories 1644430081 7282 0,7300000009437403 9 1161 2811 7086361926 9557 2,8613333364327747 43 867 2611 4702921684 8572 0,41741935476180064 5 766 2965 3977333714 11337 1,6579310161800218 19 731 1564 6290855005 5851 0,08857142925262469 2 1184 2692 3372868164 6861 0,6294999971985812 9 1077 1933 8877689391 16040 6,637419362342162 66 1112 3420 2026352035 5743 0,0063333332538604665 0 679 1553 5577150313 8304 3,113666655619939 87 754 3359 4319703577 7510 0,28733333349227924 3 760 2097 4020332650 2323 0,14699999640385303 5 1254 2428 4057192912 3838 0,05249999836087225 0 1217 1973 8053475328 14763 8,514838741671657 85 1148 2945 4445114986 4796 0,5232258109796432 6 829 2186 6962181067 10001 1,670333339770635 23 680 2017 8792009665 1919 0,02571428607084919 1 1096 2030 2022484408 11370 2,4216128837677737 36 1112 2509 8378563200 8717 2,503548401978707 58 716 3436 1927972279 916 0,09580645085342482 1 1317 2172 2320127002 4785 0,1103333327919245 1 1233 1744 5553957443 8795 1,49366666773955 23 683 1913 7007744171 11776 2,5115999865531924 32 1093 2640 1503960366 12520 2,9536666830380756 40 828 1876 8583815059 7438 0,8246666766703131 10 1261 2823 4558609924 7685 0,5493548389884728 10 1093 2033 2873212765 7555 0,6761290309650283 14 1097 1916 4388161847 10813 1,7193548346479097 23 836 3093 6775888955 2541 0,7323999975621696 11 1329 2175 1624580081 5836 0,9706666707992535 8 1270 1499 1844505072 2666 0,0086666665971279 0 1223 1603 6117666160 7046 0,12821428477764124 1 796 2261 2347167796 10077 1,1217646938474737 14 727 2139 8253242879 6842 2,337222183744112 21 1278 1887

### Calories burned vs. Activity distances and total steps

In [1]:
-- checking the relation between active distances and calories burned
SELECT TOP 20
    Id,
    Date,
    Calories,
    VeryActiveDistance, 
    ModeratelyActiveDistance, 
    LightActiveDistance
FROM dailyActivity
ORDER BY Calories DESC;

(20 rows affected)

Total execution time: 00:00:05.759

Id Date Calories VeryActiveDistance ModeratelyActiveDistance LightActiveDistance 6117666160 2016-04-21 4900 0,980000019073486 0,400000005960464 5,61999988555908 5577150313 2016-04-17 4552 5,98000001907349 0,829999983310699 2,3199999332428 8877689391 2016-04-16 4547 13,2399997711182 1,21000003814697 10,710000038147 5577150313 2016-05-01 4546 5,30999994277954 1,44000005722046 3,24000000953674 5577150313 2016-04-30 4501 5,82999992370605 0,790000021457672 2,60999989509583 8877689391 2016-04-30 4398 21,6599998474121 0,0799999982118607 4,92999982833862 5577150313 2016-04-24 4392 7,65000009536743 2,15000009536743 1,98000001907349 5577150313 2016-04-16 4274 6,6399998664856 1,27999997138977 2,73000001907349 8378563200 2016-04-21 4236 6,90000009536743 0,819999992847443 4,28999996185303 8378563200 2016-04-14 4163 5,61999988555908 1,02999997138977 3,91000008583069 8378563200 2016-04-18 4157 5,05000019073486 0,560000002384186 5,19999980926514 8877689391 2016-05-07 4142 0,0799999982118607 0,959999978542328 6,98999977111816 8378563200 2016-04-19 4092 5,30000019073486 0,879999995231628 4,17999982833862 8378563200 2016-04-13 4079 4,96000003814697 0,649999976158142 4,21000003814697 8378563200 2016-04-22 4044 4,90999984741211 0,589999973773956 4,17999982833862 4388161847 2016-05-07 4022 9,44999980926514 2,76999998092651 5,32999992370605 5577150313 2016-04-14 4022 3,32999992370605 0,310000002384186 2,77999997138977 5577150313 2016-04-21 4018 3,65000009536743 1,6599999666214 2,77999997138977 8877689391 2016-05-06 4015 12,789999961853 0,28999999165535 6,15999984741211 5577150313 2016-04-15 4005 3,92000007629395 1,60000002384186 3,55999994277954

### Calories burned vs. Activity minutes

In [2]:
-- checking a relation between activity minutes and calories burned
SELECT TOP 20 -- top 20 only for exploratory and memory saving
    Id,
    Date,
    Calories,
    VeryActiveMinutes, 
    FairlyActiveMinutes, 
    LightlyActiveMinutes,
    SedentaryMinutes
FROM dailyActivity
ORDER BY Calories DESC;

(20 rows affected)

Total execution time: 00:00:00.066

Id Date Calories VeryActiveMinutes FairlyActiveMinutes LightlyActiveMinutes SedentaryMinutes 6117666160 2016-04-21 4900 11 19 294 579 5577150313 2016-04-17 4552 200 37 159 525 8877689391 2016-04-16 4547 94 29 429 888 5577150313 2016-05-01 4546 194 72 178 499 5577150313 2016-04-30 4501 207 45 163 621 8877689391 2016-04-30 4398 124 4 223 1089 5577150313 2016-04-24 4392 210 65 141 425 5577150313 2016-04-16 4274 184 56 158 472 8378563200 2016-04-21 4236 137 16 145 677 8378563200 2016-04-14 4163 123 21 174 699 8378563200 2016-04-18 4157 117 10 174 720 8877689391 2016-05-07 4142 105 28 271 1036 8378563200 2016-04-19 4092 120 19 154 737 8378563200 2016-04-13 4079 116 14 169 680 8378563200 2016-04-22 4044 113 12 159 769 4388161847 2016-05-07 4022 120 56 260 508 5577150313 2016-04-14 4022 118 30 176 662 5577150313 2016-04-21 4018 110 74 175 670 8877689391 2016-05-06 4015 96 17 232 1095 5577150313 2016-04-15 4005 115 54 199 695

### Calories level stats

Let's see the proportion between normal calories burned and low calories burned in the tracking period.

In [57]:
SELECT 
    CaloriesLevel,
    COUNT(*) AS Frequency
FROM dailyActivity
GROUP BY CaloriesLevel;

(2 rows affected)

Total execution time: 00:00:00.063

CaloriesLevel Frequency Normal 571 Low 351

There are 571 records with normal calories level, and 351 - with low calories level.

## BMI (Body mass index) level stats

In [4]:
SELECT DISTINCT Id, BodyFatRank, COUNT (*) AS Frequency -- count a number of records of each body fat level
FROM weightLogInfo
GROUP BY Id, BodyFatRank;

(8 rows affected)

Total execution time: 00:00:00.044

Id BodyFatRank Frequency 1503960366 Normal 2 2873212765 Normal 2 6962181067 Normal 30 1927972279 Overweight 1 4319703577 Overweight 2 4558609924 Overweight 5 5577150313 Overweight 1 8877689391 Overweight 24

5\\8 users are overweight, but that data can be not clear, because of 33 users tested the app, not 8 only.

## SleepDay stats

### Each user's sleep averages:

In [7]:
-- average sleep time, average time in bed, average time before sleep by each user
SELECT 
    Id,
    AVG(TotalMinutesAsleep) AS AverageSleepTime,
    AVG(TotalTimeInBed) AS AverageTimeInBed,
    AVG(TimeBeforeSleep) AS AverageTimeBeforeSleep
FROM sleepDay
GROUP BY Id
ORDER BY AverageSleepTime;

(24 rows affected)

Total execution time: 00:00:00.119

Id AverageSleepTime AverageTimeInBed AverageTimeBeforeSleep 2320127002 61 69 8 7007744171 68 71 3 4558609924 127 140 12 3977333714 293 461 167 1644430081 294 346 52 8053475328 297 301 4 4020332650 349 379 30 6775888955 349 369 19 1503960366 360 383 22 4445114986 385 416 31 4388161847 400 423 23 4702921684 417 438 20 1927972279 417 437 20 5577150313 432 460 28 8792009665 435 453 18 8378563200 445 485 40 2347167796 446 491 44 6962181067 448 466 18 7086361926 453 466 13 5553957443 463 505 42 4319703577 476 501 25 6117666160 478 510 31 2026352035 506 537 31 1844505072 652 961 309

### Sleep quality stats:

In [21]:
-- total sleep quality stats
SELECT 
    SleepRate,
    COUNT (*) AS Frequency
FROM sleepDay
GROUP BY SleepRate;

(3 rows affected)

Total execution time: 00:00:00.351

SleepRate Frequency Lack of sleep 181 Normal sleep 190 Oversleeping 39

There are 190 records of normal sleep, 181 - of the lack of sleep, and 39 - of oversleeping.

In [3]:
-- sleep quality stats by each user
SELECT TOP 20
    Id,
    SleepRate,
    COUNT(*) AS Frequency
FROM sleepDay
GROUP BY Id, SleepRate
ORDER BY Id;

(20 rows affected)

Total execution time: 00:00:00.216

Id SleepRate Frequency 1503960366 Lack of sleep 22 1503960366 Normal sleep 1 1503960366 Oversleeping 2 1644430081 Lack of sleep 3 1644430081 Oversleeping 1 1844505072 Oversleeping 3 1927972279 Lack of sleep 3 1927972279 Normal sleep 1 1927972279 Oversleeping 1 2026352035 Lack of sleep 1 2026352035 Normal sleep 23 2026352035 Oversleeping 4 2320127002 Lack of sleep 1 2347167796 Lack of sleep 4 2347167796 Normal sleep 10 2347167796 Oversleeping 1 3977333714 Lack of sleep 27 3977333714 Normal sleep 1 4020332650 Lack of sleep 5 4020332650 Normal sleep 3

## HeartrateSeconds stats

### Average heartrates by each user:

In [9]:
SELECT 
    Id, CAST(AVG(Heartrate) AS INTEGER) AS AverageHeartrate
FROM heartrateMinutes
GROUP BY Id
ORDER BY AverageHeartrate;

(14 rows affected)

Total execution time: 00:00:00.219

Id AverageHeartrate 4388161847 63 5553957443 64 5577150313 65 8792009665 70 2347167796 73 6962181067 73 8877689391 74 2022484408 78 4558609924 78 4020332650 81 6117666160 81 7007744171 89 6775888955 89 2026352035 92

### Average daily heartrate stats by each user:

In [4]:
-- daily average heartrate statistics of each user
SELECT TOP 20
    Id, Date, CAST(AVG(Heartrate) AS INTEGER) AS AverageHeartrate
FROM heartrateMinutes
GROUP BY Id, Date
ORDER BY Id, Date;

(20 rows affected)

Total execution time: 00:00:00.671

Id Date AverageHeartrate 2022484408 2016-04-12 73 2022484408 2016-04-13 78 2022484408 2016-04-14 70 2022484408 2016-04-15 78 2022484408 2016-04-16 74 2022484408 2016-04-17 82 2022484408 2016-04-18 80 2022484408 2016-04-19 80 2022484408 2016-04-20 81 2022484408 2016-04-21 85 2022484408 2016-04-22 77 2022484408 2016-04-23 70 2022484408 2016-04-24 80 2022484408 2016-04-25 79 2022484408 2016-04-26 81 2022484408 2016-04-27 79 2022484408 2016-04-28 69 2022484408 2016-04-29 81 2022484408 2016-04-30 75 2022484408 2016-05-01 71

### HeartrateLevel by each user:

In [5]:
-- daily heartrate level stats by each user
SELECT TOP 20
    Id, 
    Date,
    HeartrateLevel,
    COUNT(*) AS Frequency
FROM heartrateMinutes
GROUP BY Id, Date, HeartrateLevel
ORDER BY Id, Date;

(20 rows affected)

Total execution time: 00:00:00.625

Id Date HeartrateLevel Frequency 2022484408 2016-04-12 Normal 562 2022484408 2016-04-12 High 62 2022484408 2016-04-12 Low 94 2022484408 2016-04-13 Normal 619 2022484408 2016-04-13 High 82 2022484408 2016-04-13 Low 55 2022484408 2016-04-14 Normal 596 2022484408 2016-04-14 Low 156 2022484408 2016-04-14 High 18 2022484408 2016-04-15 Normal 630 2022484408 2016-04-15 High 56 2022484408 2016-04-15 Low 24 2022484408 2016-04-16 High 47 2022484408 2016-04-16 Normal 321 2022484408 2016-04-16 Low 61 2022484408 2016-04-17 Low 25 2022484408 2016-04-17 Normal 603 2022484408 2016-04-17 High 75 2022484408 2016-04-18 High 81 2022484408 2016-04-18 Normal 674

In [7]:
-- total  heartrate stats by each user
SELECT TOP 20
    Id, 
    HeartrateLevel,
    COUNT(*) AS Frequency
FROM heartrateMinutes
GROUP BY Id, HeartrateLevel
ORDER BY Id;

(20 rows affected)

Total execution time: 00:00:00.158

Id HeartrateLevel Frequency 2022484408 High 2081 2022484408 Normal 17505 2022484408 Low 1496 2026352035 Normal 306 2026352035 High 118 2347167796 Low 1112 2347167796 High 1117 2347167796 Normal 20276 4020332650 Low 936 4020332650 Normal 9238 4020332650 High 1039 4388161847 Normal 19981 4388161847 Low 17407 4388161847 High 721 4558609924 Low 999 4558609924 High 1101 4558609924 Normal 24443 5553957443 Normal 27486 5553957443 High 605 5553957443 Low 12009

### Heartrate level stats during a day

In [8]:
-- this table shows how the heartrate level changes during a day 
SELECT TOP 20
    Hour,
    HeartrateLevel,
    COUNT(*) AS Frequency
FROM heartrateMinutes
GROUP BY Hour, HeartrateLevel
ORDER BY Hour;

(20 rows affected)

Total execution time: 00:00:00.234

Hour HeartrateLevel Frequency 0 Normal 6672 0 Low 3724 0 High 112 1 Low 3730 1 High 130 1 Normal 6325 2 Low 3862 2 Normal 5852 2 High 86 3 Low 4240 3 High 15 3 Normal 5551 4 Normal 5116 4 Low 4491 4 High 20 5 High 181 5 Normal 5527 5 Low 4204 6 Low 2411 6 High 838

## Sedentary and active time vs. Sleep Quality

I used to find a relation between how much time do people spend sedentary and how much time they sleep then.

In [9]:
-- a relation between sedentary minutes and sleep time, and between total steps made and sleep time
SELECT TOP 20
    SedentaryMinutes, 
    TotalSteps,
    sleepDay.TotalMinutesAsleep
FROM dailyActivity
INNER JOIN sleepDay
ON (dailyActivity.Id = sleepDay.Id AND
dailyActivity.Date = sleepDay.Date)
ORDER BY SedentaryMinutes DESC;

(20 rows affected)

Total execution time: 00:00:00.204

SedentaryMinutes TotalSteps TotalMinutesAsleep 1265 3702 59 1222 1982 77 1219 2268 98 1193 3176 119 1167 1675 166 1155 4631 79 1153 7550 74 1142 5600 58 1129 5079 61 1121 3428 115 1106 4676 106 1096 7833 171 1076 19769 74 1062 9148 103 1058 3761 296 1053 838 437 1053 4053 235 1020 10218 62 1019 6543 123 1017 11009 99

## BMI vs. SedentaryMinutes & TotalSteps

In [10]:
-- a relation between sedentary time, total steps made and body mass index
SELECT TOP 20
    SedentaryMinutes,
    TotalSteps,
    weightLogInfo.BodyMassIndex
FROM dailyActivity
INNER JOIN weightLogInfo ON (dailyActivity.Id = weightLogInfo.Id
AND dailyActivity.Date = weightLogInfo.Date)
ORDER BY BodyMassIndex;

(20 rows affected)

Total execution time: 00:00:00.058

SedentaryMinutes TotalSteps BodyMassIndex 1057 8859 21,4500007629395 720 7566 21,6900005340576 816 15103 22,6499996185303 798 14727 22,6499996185303 731 10081 23,8199996948242 654 12109 23,8199996948242 683 10147 23,8500003814697 689 11404 23,8899993896484 697 10320 23,8899993896484 621 12627 23,8899993896484 745 6815 23,8899993896484 709 4188 23,8899993896484 591 10524 23,9300003051758 659 10742 23,9599990844727 634 10145 23,9599990844727 708 11835 23,9599990844727 659 10725 23,9599990844727 645 10762 23,9599990844727 837 5563 24 484 20031 24

## Sedentary Minutes vs. Heartrate levels

In [11]:
-- a relation between how much time people spend sedentary and how it affects their heart
WITH dailyHeartrate AS
(
    SELECT TOP 20
        Id, 
        Date,
        HeartrateLevel,
        COUNT(*) AS Frequency
FROM heartrateMinutes
GROUP BY Id, Date, HeartrateLevel
)
SELECT 
    SedentaryMinutes,
    dailyHeartrate.HeartrateLevel,
    dailyHeartrate.Frequency
FROM dailyActivity
INNER JOIN dailyHeartrate ON
(dailyActivity.Id = dailyHeartrate.Id AND
 dailyActivity.Date = dailyHeartrate.Date)
ORDER BY SedentaryMinutes DESC;

(20 rows affected)

Total execution time: 00:00:00.224

SedentaryMinutes HeartrateLevel Frequency 1438 Normal 2 1318 High 129 1240 Normal 807 1238 Normal 620 1152 Low 22 1123 High 8 1119 High 57 1048 Low 73 1040 High 48 858 Low 858 764 Low 5 732 High 51 696 Low 597 670 Normal 685 662 High 144 659 High 116 653 Normal 1297 623 Low 137 623 Normal 1112 508 Low 752